[CIR][X86] Add support for `cpuid`/`cpuidex`

rturrado opened the pull request:
Adds support for the `__builtin_ia32_cpuid` and `__builtin_ia32_cpuidex` X86 builtins.

Part of [167765](https://github.com/llvm/llvm-project/issues/167765).

rturrado committed:
[CIR][X86] Add support for cpuid/cpuidex

Lancern commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "clang/CIR/Dialect/IR/CIRTypes.h"
 #include "clang/CIR/MissingFeatures.h"
 #include "llvm/ADT/Sequence.h"
+#include "llvm/IR/InlineAsm.h"
```
Why is this needed?

rturrado replied:
Good spot. This comes from a first implementation. Now removed.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -5819,4 +5762,31 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+          CIR_SInt32:$funcId, CIR_SInt32:$subFuncId);
+  // TODO: remove once we can return an optional mlir::Value from
+  //   emitX86BuiltinExpr
+  let results = (outs CIR_VectorType:$result);
```
You're not returning the result value from `emitX86BuiltinExpr` so the results here are unnecessary.

Lancern commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -4248,6 +4193,76 @@ mlir::LogicalResult CIRToLLVMAwaitOpLowering::matchAndRewrite(
+    return mlir::failure();
+  if (opNameAttr.getValue() == "cpuid")
+    subFuncId = mlir::LLVM::ConstantOp::create(rewriter, op.getLoc(), i32Ty, 0);
+  std::vector operands{funcId, subFuncId};
```
Use an array here instead.

rturrado replied:
You mean an `ArrayRef<mlir::Value>` right?

Lancern commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -4248,6 +4193,76 @@ mlir::LogicalResult CIRToLLVMAwaitOpLowering::matchAndRewrite(
+      moduleOp->getAttrOfType<mlir::StringAttr>("llvm.target_triple");
+  if (!tripleAttr)
+    return mlir::failure();
+  llvm::Triple triple(tripleAttr.getValue().str());
```
A more reliable way to get the target triple would be through the `lowerMod` member attached to every `CIRToLLVM` lowering patterns:

```cpp
const llvm::Triple &triple = lowerMod->getTarget().getTriple();
```

rturrado commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -1835,7 +1836,16 @@ CIRGenFunction::emitX86BuiltinExpr(unsigned builtinID, const CallExpr *expr) {
   case X86::BI__builtin_ia32_cvtneps2bf16_256_mask:
   case X86::BI__builtin_ia32_cvtneps2bf16_512_mask:
```
Good point, thanks!

rturrado commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -5819,4 +5762,31 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+  let description = [{
+    The `cir.cpuid` operation takes a base pointer to an array of 4 integers, a
+    function ID and a sub-function ID. The array of 4 integers is filled with
+    different information about the processor.
```
Is this just a `let hasVerifier = 1;` line or something else?

rturrado replied:
I have updated the `argument` variable to:
```
let arguments =
      (ins Arg<CIR_PtrToI32Array4, "array address", [MemWrite]>:$basePtr,
          CIR_SInt32:$funcId, CIR_SInt32:$subFuncId);
```

rturrado replied:
Do you think we have to set `let hasVerifier = 1` and implement `verify()` to add extra checks on `basePtr`?

rturrado replied:
It seems `CIR_PtrToI32Array4` does not exist. So I will try to proceed via the `hasVerifier` option.

rturrado commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
````diff
@@ -5819,4 +5762,31 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+    Example:
+
+    ```mlir
+    cir.cpuid %basePtr, %funcId, %subFuncId
````
I have added a little description for each argument.

rturrado commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -4248,6 +4193,76 @@ mlir::LogicalResult CIRToLLVMAwaitOpLowering::matchAndRewrite(
+
+  mlir::Value funcId = adaptor.getFuncId();
+  mlir::Value subFuncId = adaptor.getSubFuncId();
+  mlir::StringAttr opNameAttr = op->getAttrOfType<mlir::StringAttr>("name");
```
I've removed that code section. `subFuncId` is always built correctly, either for `cpuid` or for `cpuidex`, at CIRGenBuiltinX86.cpp.

rturrado committed:
Address review comments from Lancern

rturrado committed:
Fix cpuid/cpuidex builtin implementation

These builtins do not return anything, so the first parameter is the first operand.

rturrado committed:
Change CIR_CpuIdOp to accept a pointer as first operand

Arrays are usually decayed to pointers. By not restricting basePtr to be exactly a pointer to an array of 4 signed integers, we allow passing a pointer to signed integer. Extra verifications are already done at the verifier.

rturrado committed:
Add cpuid-builtins.c test

rturrado committed:
Add LLVM and OGCG checks to cpuid-builtins.c test

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -5896,4 +5896,39 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+  }];
+
+  let arguments =
+      (ins Arg<CIR_PointerType, "array address", [MemWrite]>:$basePtr,
```
```suggestion
      (ins Arg<CIR_PtrToArray, "array address", [MemWrite]>:$basePtr,
```

andykaylor commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -3565,6 +3565,31 @@ cir::EhTypeIdOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+    return emitOpError()
+           << "base pointer must point to an array of size at least 4";
+
+  // Array decay: basePtr points to !s32i
```
This is not necessarily an array decay. The user can pass any arbitrary pointer to integer to this function. For example: 

```
  int *cpuInfo = malloc(4 * sizeof(int));
  __cpuid((int *)&cpuInfo, functionId);
```

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -5896,4 +5896,39 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+  }];
+
+  let arguments =
+      (ins Arg<CIR_PointerType, "array address", [MemWrite]>:$basePtr,
```
```suggestion
      (ins Arg<CIR_PtrTo<CIR_SInt32>, "array address", [MemWrite]>:$basePtr,
```
I think if you do this you don't need the verifier.

rturrado replied:
I'll give it a go!

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/cpuid-builtins.c:
```diff
@@ -0,0 +1,148 @@
+// CIR-LABEL: __cpuid_with_array_decayed_to_pointer
+// CIR: %[[CPUINFO_PTR:.*]] = cir.alloca !cir.ptr<!s32i>
+// CIR: %[[ZERO:.*]] = cir.const #cir.int<0> : !s32i
+// CIR: {{.*}}cir.cpuid{{.*}}(%{{.*}}, %{{.*}}, %[[ZERO]]) : (!cir.ptr<!s32i>, !s32i, !s32i)
```
What are the `{{.*}}` before and after `cir.cpuid` masking? This doesn't appear to match the example in the .td file.

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/cpuid-builtins.c:
```diff
@@ -0,0 +1,148 @@
+// CIR: {{.*}}cir.cpuid{{.*}}(%{{.*}}, %{{.*}}, %[[ZERO]]) : (!cir.ptr<!s32i>, !s32i, !s32i)
+
+// LLVM-LABEL: __cpuid_with_array_decayed_to_pointer
+// LLVM-DAG: [[STACK_PTR_TO_CPU_INFO_PTR:%.*]] = alloca ptr
```
Why are you using `DAG` here?

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/cpuid-builtins.c:
```diff
@@ -0,0 +1,148 @@
+// LLVM-DAG: [[STACK_PTR_TO_SUBFUNCTION_ID:%.*]] = alloca i32
+// LLVM-DAG: [[CPU_INFO_PTR:%.*]] = load ptr, ptr [[STACK_PTR_TO_CPU_INFO_PTR]]
+// LLVM-DAG: [[FUNCTION_ID:%.*]] = load i32, ptr [[STACK_PTR_TO_FUNCTION_ID]]
+// LLVM-DAG: [[SUBFUNCTION_ID:%.*]] = load i32, ptr [[STACK_PTR_TO_SUBFUNCTION_ID]]
```
```suggestion
// LLVM: [[CPU_INFO_PTR:%.*]] = load ptr
// LLVM: [[FUNCTION_ID:%.*]] = load i32
// LLVM: [[SUBFUNCTION_ID:%.*]] = load i32
```

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/cpuid-builtins.c:
```diff
@@ -0,0 +1,148 @@
+// LLVM-DAG: store i32 [[RESULT3]], ptr [[ADDRPTR3]], align 4
+
+// OGCG-LABEL: __cpuidex
+// OGCG-DAG: [[STACK_PTR_TO_CPU_INFO_PTR:%.*]] = alloca ptr
```
Why aren't you matching the loads here and using them in the checks below as you did above? The allocas don't test anything useful.

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/cpuid-builtins.c:
```diff
@@ -0,0 +1,148 @@
+// OGCG-DAG: [[STACK_PTR_TO_FUNCTION_ID:%.*]] = alloca i32
+// OGCG-DAG: [[STACK_PTR_TO_SUBFUNCTION_ID:%.*]] = alloca i32
+// OGCG-DAG: [[ASMRESULTS:%.*]] = call { i32, i32, i32, i32 } asm "{{.*}}cpuid{{.*}}", "={ax},=r,={cx},={dx},0,2"(i32 %{{.*}}, i32 %{{.*}})
+// OGCG-DAG: [[RESULT0:%.*]] = extractvalue { i32, i32, i32, i32 } %{{.*}}, 0
```
Using `ASMRESULTS` in these checks would be good.

rturrado commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -3565,6 +3565,31 @@ cir::EhTypeIdOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+
+  // basePtr points to an array of size at least 4
+  auto arrayTy = mlir::dyn_cast<cir::ArrayType>(type);
+  if (arrayTy && (arrayTy.getSize() < 4))
```
I think you are right, yes. For the 3 tests in the current `ms-intrinsics-cpuid.c` test file, the emitted CIR looks like:

```
"cir.cpuid"(%{{.*}}, %{{.*}}, %{{.*}}) : (!cir.ptr<!s32i>, !s32i, !s32i) -> () loc(#loc{{.*}})
```

rturrado committed:
Address comments from Andy Kaylor's review

CIRDialect.cpp: remove CpuIdOp::verify().
CIROps.td: change type of cpu_info parameter to CIR_PtrToType<CIR_SInt32>.
cpuid-builtins.c: remove alloca checks, change come LLVM-DAG and OGCG-DAG checks to LLVM and OGCG respectively, use variable names whenever possible.
LowerToLLVM.cpp: do not use lowerMod to get the triple.

rturrado committed:
Fix hint names for builtin arguments

rturrado committed:
Fix tests

andykaylor commented on clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp:
```diff
@@ -4282,6 +4283,71 @@ mlir::LogicalResult CIRToLLVMAwaitOpLowering::matchAndRewrite(
+  std::array<mlir::Value, 2> operands{funcId, subFuncId};
+
+  StringRef asmString, constraints;
+  mlir::ModuleOp module = op->getParentOfType<mlir::ModuleOp>();
```
```suggestion
  mlir::ModuleOp mod = op->getParentOfType<mlir::ModuleOp>();
```
`module` is a keyword in C++ 20. It's best to avoid using it as an identifier name (though I see we do that in a lot of CIR code, which should also be fixed).

rturrado committed:
Rename cpuid-builtins.c to ms-x86-intrinsics.c

Add -ffreestanding flag.
Move checks to inside each function.

rturrado committed:
Address comments from Andy Kaylor's review

CIROps.td: fit lines to 80 columns.
LowerToLLVM.cpp: rename module to moduleOp.
ms-x86-intrinsics.c: update cir.cpuid checks to new output.

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -5896,4 +5896,38 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+    stores it in an array of 4 integers.
+
+    This operation takes 3 arguments: `cpu_info`, a pointer to an array of 4
+    integers; `function_id`, an integer determining what type of information to be
```
Unfortunately, no.

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -5896,4 +5896,44 @@ def CIR_BlockAddressOp : CIR_Op<"block_address", [Pure]> {
+        qualified(type($cpu_info)) `,`
+        type($function_id) `,`
+        type($sub_function_id) attr-dict
+    }];
```
```suggestion
  let assemblyFormat = [{
      $cpu_info`,` $function_id`,` $sub_function_id `:`
          qualified(type($cpu_info)) `,`
          type($function_id) `,`
          type($sub_function_id) attr-dict
      }];
```
Nit

rturrado replied:
Are you sure about this? The rest of the file seems to use an indentation of 2 spaces.

andykaylor reviewed (commented):
Sorry, just a couple of last nits

rturrado committed:
Address comments from Andy Kaylor's review

CIROps.td: use indentation of 2 spaces, and format arguments similar to how it is done in the rest of the file.
LowerToLLVM.cpp: use {functionId, subFunctionId} instead of operands.
ms-x86-intrinsics.c: change all LLVM-DAG and OGCG-DAG checks to LLVM and OGCG respectively.

andykaylor reviewed (approved):
lgtm

rturrado commented:
Many thanks for the thorough and smooth reviews, @Lancern and @andykaylor!
